fastlane/fastlane
# [upload_app_privacy_details_to_app_store] fix error when `Info.plist` contains `NSUserTrackingUsageDescription`

### Checklist

### Motivation and Context

Resolves #17895
Resolves #19487
Resolves #19781

### Description

The problem was well described in the associated tickets, but basically: Apple does server side validation when attempting to delete `ADVERTISING_DATA` and the app's Info.plist contains `NSUserTrackingUsageDescription` 😬 So we can't delete it.

### Design Decisions

I thought of adding an option like this:

```ruby
FastlaneCore::ConfigItem.new(key: :skip_advertising_data_deletion_if_missing,
                             env_name: "UPLOAD_APP_PRIVACY_DETAILS_TO_APP_STORE_SKIP_DELETION_ADVERTISING_DATA",
                             description: "Used to skip the deletion of the advertising data app privacy settings when such setting is present in App Store Connect, but not locally. It should rarely be used",
                             type: Boolean,
                             default_value: false),
```

But I ended up realizing that it probably would be too useful. People changing these configs are not something common, specially the advertisement one, probably. So when they do, it's a one-off, so they can do it manually via App Store Connect (like I explained in the `UI.important` messages - please review them 🤗 )

### Fun Facts

When the data is something like:

```json
{
    "category": "ADVERTISING_DATA",
    "purposes": [
        "THIRD_PARTY_ADVERTISING"
    ],
    "data_protections": [
        "DATA_NOT_LINKED_TO_YOU",
        "DATA_USED_TO_TRACK_YOU"
    ]
}
```

The API returns 2 elements for `ADVERTISING_DATA` (`DATA_NOT_LINKED_TO_YOU` and `DATA_USED_TO_TRACK_YOU`). If you attempt to delete each of them, both will fail (HTTP 409):

```json
{
  "errors": [
    {
      "id": "b47e8827-0e75-4a1c-ace8-b4f27b9d37e0",
      "status": "409",
      "code": "STATE_ERROR.BINARY_INDICATES_APP_TRACKS_USERS",
      "title": "Your binary indicates that your app tracks users.",
      "detail": "Your app indicates that you track users."
    }
  ]
}
```

However, the `DATA_NOT_LINKED_TO_YOU` actually gets deleted although it errors out, whereas the `DATA_USED_TO_TRACK_YOU` really can't be deleted. 🤷🤷🤷

### Testing Steps

To test this branch, modify your Gemfile as:

```ruby
gem 'fastlane', git: 'https://github.com/fastlane/fastlane.git', branch: 'rogerluan-fix-upload-app-privacy-details-when-tracking-advertising-data'
```

And run `bundle install` to apply the changes.

Tagging people who were having issues with this:

- @gavrichards
- @khushbujilka22
- @mortenholmgaard
- @PWhittle86
- @sidepelican
- @tguidon
- @wimbledon
- @yood

Perhaps someone could help me triple check if my fix is alright? 🙇
